File: backend/src/ratatoskChannel.ts
import { existsSync, readFileSync, writeFileSync } from "node:fs";
import { join } from "node:path";
import { hasOwnRatatoskAccount, ownRatatoskEmail, getOwnV2Session } from "./ratatoskOwnAccount.js";
import { listConversations, getRecentMessages, sendPresenceHeartbeat, getServerNow, type RatatoskConversation } from "./ratatosk.js";

// No push from Ratatosk -- its own UI polls every 4-5s; this owner-DM
// control channel is far less latency-sensitive (it's "give Caroline an
// instruction", not a live conversation someone's staring at), so a longer
// interval is fine and keeps this from hammering the backend.
const POLL_INTERVAL_MS = 15_000;

// Deliberately its OWN, much faster interval -- NOT piggybacked on
// POLL_INTERVAL_MS above. portal/chat.js's presence TTL is 15s; reusing the
// 15s owner-DM poll for the heartbeat too would put every heartbeat right at
// the TTL boundary, so any tick that ran even slightly late (network hiccup,
// event-loop delay) would let Caroline visibly flicker offline. 5s matches
// chat.js's own PRESENCE_POLL_MS exactly, giving 3 heartbeats per TTL window.
const PRESENCE_INTERVAL_MS = 5_000;

// Keyed by groupId -- one cursor per monitored group, not a single global
// value, since (per explicit instruction, 2026-09-03) this channel now
// watches every group Caroline is a member of, not just one hardcoded
// owner DM. Replaces the old ratatosk-owner-dm-cursor.json (single-group)
// format entirely -- a fresh file, not a migration, since that old cursor's
// one timestamp was itself tied to whichever single group used to be
// hardcoded and has no meaningful mapping onto "per group" here.
function cursorsPath(workspaceDir: string): string {
  return join(workspaceDir, "ratatosk-groups-cursor.json");
}

function loadCursors(workspaceDir: string): Record<string, number> {
  try {
    if (!existsSync(cursorsPath(workspaceDir))) return {};
    return JSON.parse(readFileSync(cursorsPath(workspaceDir), "utf-8")) as Record<string, number>;
  } catch (err) {
    console.error("[caroline] [ratatosk-channel] loadCursors failed (treating as never-checked):", err);
    return {};
  }
}

function saveCursors(workspaceDir: string, cursors: Record<string, number>): void {
  try {
    writeFileSync(cursorsPath(workspaceDir), JSON.stringify(cursors, null, 2) + "\n", "utf-8");
  } catch (err) {
    console.error("[caroline] [ratatosk-channel] saveCursors failed (best-effort, next tick re-derives it):", err);
  }
}

/**
 * In-memory snapshot of the channel's own state, for the ratatosk_status_get
 * control op (Settings/external tooling) -- confirmed live (2026-08-31) on a
 * DIFFERENT subsystem (AppBrowserHost) that a poll loop with no introspection
 * and thin logging is nearly impossible to diagnose after the fact; this is
 * the same fix applied here from the start instead of after an incident.
 */
interface ChannelStatus {
  enabled: boolean;
  tickCount: number;
  lastTickAtIso: string | null;
  lastTickOutcome: string | null;
  carolineEmail: string | null;
  monitoredGroupCount: number;
  lastError: string | null;
  lastErrorAtIso: string | null;
}

const status: ChannelStatus = {
  enabled: false,
  tickCount: 0,
  lastTickAtIso: null,
  lastTickOutcome: null,
  carolineEmail: null,
  monitoredGroupCount: 0,
  lastError: null,
  lastErrorAtIso: null,
};

export function getRatatoskChannelStatus(): ChannelStatus {
  return { ...status };
}

/**
 * Poll loop watching EVERY group Caroline's own Ratatosk account is a member
 * of (not just one hardcoded owner DM -- per explicit correction, 2026-09-03:
 * "Кэролайн должна мониторить ВСЕ группы в которых состоит" -- Caroline
 * previously only ever watched a single DM with the owner, resolved once and
 * cached; being added to any OTHER group, or the owner's phone using a
 * different/duplicate DM than the one cached in memory, meant new messages
 * there were simply never seen, with no error or symptom beyond silence) for
 * new messages from anyone other than herself, injecting each as a proactive
 * turn into the headless "ratatosk" session (see server.ts, which owns
 * creating that session and passes injectFromOwner as its injectProactive).
 *
 * A complete no-op (checked, logged, does nothing further) until Caroline has
 * her own Ratatosk account -- checked on every tick, not just once at
 * startup, so the channel starts working the moment that becomes true
 * without needing a restart. No longer requires the owner to be logged in on
 * THIS machine either: this channel is about what Caroline's own account
 * sees, independent of that. Every tick logs an unconditional heartbeat
 * (what it saw, what it decided) -- same reasoning as server.ts's
 * checkHang(): better a noisy log than another "I can't tell what happened".
 */
export function startRatatoskOwnerChannel(workspaceDir: string, injectFromOwner: (text: string) => void): void {
  console.error(`[caroline] [ratatosk-channel] starting poll loop (interval=${POLL_INTERVAL_MS}ms)`);

  setInterval(async () => {
    status.tickCount++;
    status.lastTickAtIso = new Date().toISOString();
    const tickLabel = `[ratatosk-channel] tick #${status.tickCount}`;
    try {
      const hasOwn = hasOwnRatatoskAccount(workspaceDir);
      status.enabled = hasOwn;
      if (!hasOwn) {
        status.lastTickOutcome = "skipped (no own Ratatosk account yet)";
        console.error(`[caroline] ${tickLabel}: ${status.lastTickOutcome}`);
        return;
      }
      const carolineEmail = ownRatatoskEmail(workspaceDir)!;
      status.carolineEmail = carolineEmail;

      console.error(`[caroline] ${tickLabel}: minting Caroline's own v2 session (email=${carolineEmail})...`);
      const session = await getOwnV2Session(workspaceDir);

      const conversations = await listConversations(session, carolineEmail);
      status.monitoredGroupCount = conversations.length;
      console.error(`[caroline] ${tickLabel}: monitoring ${conversations.length} group(s)`);

      const cursors = loadCursors(workspaceDir);
      const perGroupNew: { group: RatatoskConversation; texts: string[] }[] = [];
      let totalNew = 0;

      for (const group of conversations) {
        const messages = await getRecentMessages(session, group.groupId, 2);
        if (messages.length === 0) continue;
        const maxTs = Math.max(...messages.map((m) => m.ts));

        // On a genuinely first-ever check of a given group (no cursor entry yet --
        // fresh workspace, or Caroline was just added to this group), treat "already
        // seen" as 24h ago rather than "everything up to right now" -- confirmed live
        // (2026-09-02) as a real bug on the single-group predecessor of this loop: the
        // OLD behavior (seed cursor to maxTs, reply to nothing) silently swallowed
        // real messages that had just arrived, because a brand-new group's very first
        // tick treated them exactly like day-old history. Genuinely old history
        // (predating this group's first check by more than a day) still isn't
        // replayed -- only this 24h window is.
        const lastSeenTs = cursors[group.groupId] ?? (getServerNow() - 24 * 3600_000);
        cursors[group.groupId] = maxTs;

        const newMessages = messages.filter((m) => m.ts > lastSeenTs && m.from?.toLowerCase() !== carolineEmail.toLowerCase());
        if (newMessages.length === 0) continue;
        const texts = newMessages.map((m) => `${m.from ?? "unknown"}: ${m.text ?? ""}`).filter((t) => t.trim().length > 0);
        if (texts.length === 0) continue;
        perGroupNew.push({ group, texts });
        totalNew += texts.length;
      }
      saveCursors(workspaceDir, cursors);

      if (perGroupNew.length === 0) {
        status.lastTickOutcome = "no new messages in any monitored group since last cursor";
        console.error(`[caroline] ${tickLabel}: ${status.lastTickOutcome}`);
        return;
      }

      status.lastTickOutcome = `${totalNew} new message(s) across ${perGroupNew.length} group(s), injecting into headless session`;
      console.error(`[caroline] ${tickLabel}: ${status.lastTickOutcome}`);
      // Every group's exact groupId is spelled out right next to its own messages --
      // confirmed live (2026-09-03) as a real bug (on the single-group predecessor of
      // this loop) that a bare "reply there" gives the model no actual value to latch
      // onto, so it can pick a groupId from unrelated prior context/memory instead of
      // the real one, and the reply silently lands in the wrong conversation. With
      // multiple groups possibly pending at once here, that risk is even higher, so
      // this is spelled out once per group, not once for the whole batch.
      const sections = perGroupNew.map(({ group, texts }) =>
        `Group ${JSON.stringify(group.name)} (groupId="${group.groupId}") -- reply here via ratatosk_send_message ` +
          `as:"caroline" groupId:"${group.groupId}" (this exact groupId, not one from earlier in your own history/memory):\n` +
          texts.join("\n"));
      injectFromOwner(
        `[New Ratatosk message(s) since you last checked, across ${perGroupNew.length} group(s) -- reply to each ` +
          `using its OWN groupId shown below, not in any chat window:\n\n${sections.join("\n\n")}]`,
      );
    } catch (err) {
      status.lastError = err instanceof Error ? err.message : String(err);
      status.lastErrorAtIso = new Date().toISOString();
      status.lastTickOutcome = `threw: ${status.lastError}`;
      console.error(`[caroline] ${tickLabel}: threw:`, err);
    }
  }, POLL_INTERVAL_MS);
}

/**
 * Keeps Caroline showing as online in Ratatosk/SquirrelWisdom, everywhere
 * she appears (not scoped to a single group -- presence is keyed only by
 * her own email, see sendPresenceHeartbeat's sw_presence/{email} path), by
 * publishing a fresh heartbeat every PRESENCE_INTERVAL_MS. Deliberately its
 * OWN setInterval, independent of startRatatoskOwnerChannel above -- see
 * PRESENCE_INTERVAL_MS's own comment for why the two must not share a timer.
 * Only gated on Caroline having her own account; unlike the owner-DM poll
 * loop, this does NOT need the owner to be logged in on this machine --
 * presence is Caroline's own status, not a channel to the owner.
 */
export function startRatatoskPresenceHeartbeat(workspaceDir: string): void {
  console.error(`[caroline] [ratatosk-presence] starting heartbeat loop (interval=${PRESENCE_INTERVAL_MS}ms)`);
  let tickCount = 0;

  setInterval(async () => {
    tickCount++;
    try {
      if (!hasOwnRatatoskAccount(workspaceDir)) {
        if (tickCount === 1) console.error(`[caroline] [ratatosk-presence] tick #${tickCount}: skipped (no own account yet)`);
        return;
      }
      const carolineEmail = ownRatatoskEmail(workspaceDir)!;
      const session = await getOwnV2Session(workspaceDir);
      await sendPresenceHeartbeat(session, carolineEmail);
    } catch (err) {
      console.error(`[caroline] [ratatosk-presence] tick #${tickCount}: threw:`, err);
    }
  }, PRESENCE_INTERVAL_MS);
}

File: backend-py/app/ratatosk_channel.py
"""Ports backend/src/ratatoskChannel.ts -- the headless owner-DM control
channel (watches EVERY Ratatosk group Caroline's own account is a member
of for new messages from anyone else, injecting each as a proactive turn)
plus the independent presence-heartbeat loop. Pure logic + module-level
status only; wiring a headless ChatSession to actually receive the
injected text lives in main.py (mirrors the original's own split between
this module and server.ts).
"""

from __future__ import annotations

import asyncio
import json
import time
from dataclasses import asdict, dataclass
from datetime import datetime, timezone
from pathlib import Path
from typing import Any, Awaitable, Callable

from app.logging_setup import log_event
from app.plugins.ratatosk_api import get_recent_messages, get_server_now, list_conversations, send_presence_heartbeat
from app.plugins.ratatosk_own_account import get_own_v2_session, has_own_ratatosk_account, own_ratatosk_email

# No push from Ratatosk -- its own UI polls every 4-5s; this owner-DM
# control channel is far less latency-sensitive (it's "give Caroline an
# instruction", not a live conversation someone's staring at), so a longer
# interval is fine and keeps this from hammering the backend.
POLL_INTERVAL_MS = 15_000

# Deliberately its OWN, much faster interval -- NOT piggybacked on
# POLL_INTERVAL_MS above. chat.js's presence TTL is 15s; reusing the 15s
# owner-DM poll for the heartbeat too would put every heartbeat right at
# the TTL boundary, so any tick that ran even slightly late would let
# Caroline visibly flicker offline. 5s matches chat.js's own
# PRESENCE_POLL_MS exactly, giving 3 heartbeats per TTL window.
PRESENCE_INTERVAL_MS = 5_000


def _cursors_path(workspace_dir: str) -> Path:
    return Path(workspace_dir) / "ratatosk-groups-cursor.json"


def _load_cursors(workspace_dir: str) -> dict[str, float]:
    path = _cursors_path(workspace_dir)
    if not path.exists():
        return {}
    try:
        return json.loads(path.read_text(encoding="utf-8"))
    except Exception as exc:
        log_event("plugin:ratatosk-channel", "load_cursors_failed", error=str(exc))
        return {}


def _save_cursors(workspace_dir: str, cursors: dict[str, float]) -> None:
    try:
        _cursors_path(workspace_dir).write_text(json.dumps(cursors, indent=2) + "\n", encoding="utf-8")
    except Exception as exc:
        log_event("plugin:ratatosk-channel", "save_cursors_failed", error=str(exc))


@dataclass
class ChannelStatus:
    """In-memory snapshot of the channel's own state, for the
    ratatosk_channel_status control op (Settings/external tooling) -- a
    poll loop with no introspection and thin logging is nearly impossible
    to diagnose after the fact."""

    enabled: bool = False
    tick_count: int = 0
    last_tick_at_iso: str | None = None
    last_tick_outcome: str | None = None
    caroline_email: str | None = None
    monitored_group_count: int = 0
    last_error: str | None = None
    last_error_at_iso: str | None = None


_status = ChannelStatus()


def get_ratatosk_channel_status() -> dict[str, Any]:
    return asdict(_status)


async def _owner_channel_tick(workspace_dir: str, inject_from_owner: Callable[[str], None]) -> None:
    _status.tick_count += 1
    _status.last_tick_at_iso = datetime.now(timezone.utc).isoformat().replace("+00:00", "Z")
    tick_label = f"tick #{_status.tick_count}"
    try:
        has_own = has_own_ratatosk_account(workspace_dir)
        _status.enabled = has_own
        if not has_own:
            _status.last_tick_outcome = "skipped (no own Ratatosk account yet)"
            log_event("plugin:ratatosk-channel", "tick", label=tick_label, outcome=_status.last_tick_outcome)
            return
        caroline_email = own_ratatosk_email(workspace_dir)
        assert caroline_email is not None
        _status.caroline_email = caroline_email

        session = await get_own_v2_session(workspace_dir)
        conversations = await list_conversations(session, caroline_email)
        _status.monitored_group_count = len(conversations)
        log_event("plugin:ratatosk-channel", "tick_monitoring", label=tick_label, group_count=len(conversations))

        cursors = _load_cursors(workspace_dir)
        per_group_new: list[tuple[dict[str, Any], list[str]]] = []
        total_new = 0

        for group in conversations:
            group_id = group["groupId"]
            messages = await get_recent_messages(session, group_id, 2)
            if not messages:
                continue
            max_ts = max(m.get("ts", 0) for m in messages)

            # On a genuinely first-ever check of a given group (no cursor
            # entry yet -- fresh workspace, or Caroline was just added to
            # this group), treat "already seen" as 24h ago rather than
            # "everything up to right now" -- the OLD behavior (seed cursor
            # to max_ts, reply to nothing) silently swallowed real messages
            # that had just arrived. Genuinely old history (predating this
            # group's first check by more than a day) still isn't replayed.
            last_seen_ts = cursors.get(group_id, get_server_now() - 24 * 3600_000)
            cursors[group_id] = max_ts

            new_messages = [
                m for m in messages
                if m.get("ts", 0) > last_seen_ts and (m.get("from") or "").lower() != caroline_email.lower()
            ]
            if not new_messages:
                continue
            texts = [f"{m.get('from') or 'unknown'}: {m.get('text') or ''}" for m in new_messages]
            texts = [t for t in texts if t.strip()]
            if not texts:
                continue
            per_group_new.append((group, texts))
            total_new += len(texts)

        _save_cursors(workspace_dir, cursors)

        if not per_group_new:
            _status.last_tick_outcome = "no new messages in any monitored group since last cursor"
            log_event("plugin:ratatosk-channel", "tick", label=tick_label, outcome=_status.last_tick_outcome)
            return

        _status.last_tick_outcome = f"{total_new} new message(s) across {len(per_group_new)} group(s), injecting into headless session"
        log_event("plugin:ratatosk-channel", "tick", label=tick_label, outcome=_status.last_tick_outcome)

        # Every group's exact groupId is spelled out right next to its own
        # messages -- a bare "reply there" gives the model no actual value
        # to latch onto, so it can pick a groupId from unrelated prior
        # context/memory instead of the real one, and the reply silently
        # lands in the wrong conversation. Spelled out once per group, not
        # once for the whole batch, since multiple groups can be pending.
        sections = [
            f'Group "{group["name"]}" (groupId="{group["groupId"]}") -- reply here via ratatosk_send_message '
            f'as:"caroline" groupId:"{group["groupId"]}" (this exact groupId, not one from earlier in your own '
            f"history/memory):\n" + "\n".join(texts)
            for group, texts in per_group_new
        ]
        inject_from_owner(
            f"[New Ratatosk message(s) since you last checked, across {len(per_group_new)} group(s) -- reply to "
            f"each using its OWN groupId shown below, not in any chat window:\n\n" + "\n\n".join(sections) + "]"
        )
    except Exception as exc:
        _status.last_error = str(exc)
        _status.last_error_at_iso = datetime.now(timezone.utc).isoformat().replace("+00:00", "Z")
        _status.last_tick_outcome = f"threw: {_status.last_error}"
        log_event("plugin:ratatosk-channel", "tick_failed", label=tick_label, error=_status.last_error)


def start_ratatosk_owner_channel(workspace_dir: str, inject_from_owner: Callable[[str], None]) -> asyncio.Task[None]:
    log_event("plugin:ratatosk-channel", "starting_poll_loop", interval_ms=POLL_INTERVAL_MS)

    async def _loop() -> None:
        while True:
            await asyncio.sleep(POLL_INTERVAL_MS / 1000)
            await _owner_channel_tick(workspace_dir, inject_from_owner)

    return asyncio.create_task(_loop())


def start_ratatosk_presence_heartbeat(workspace_dir: str) -> asyncio.Task[None]:
    log_event("plugin:ratatosk-presence", "starting_heartbeat_loop", interval_ms=PRESENCE_INTERVAL_MS)

    async def _loop() -> None:
        tick_count = 0
        while True:
            await asyncio.sleep(PRESENCE_INTERVAL_MS / 1000)
            tick_count += 1
            try:
                if not has_own_ratatosk_account(workspace_dir):
                    if tick_count == 1:
                        log_event("plugin:ratatosk-presence", "tick_skipped_no_account", tick=tick_count)
                    continue
                caroline_email = own_ratatosk_email(workspace_dir)
                assert caroline_email is not None
                session = await get_own_v2_session(workspace_dir)
                await send_presence_heartbeat(session, caroline_email)
            except Exception as exc:
                log_event("plugin:ratatosk-presence", "tick_failed", tick=tick_count, error=str(exc))

    return asyncio.create_task(_loop())
